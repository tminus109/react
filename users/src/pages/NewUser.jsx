import { Link } from "react-router-dom";
import Header from "../components/Header";
import UserForm from "../components/UserForm";

const headerProps = {
  title: "New User",
};

const userFormProps = {
  legend: "Add new user",
  btnType: "submit",
  btnText: "Add user",
  user: {
    first_name: "",
    last_name: "",
    status: "active",
  },
  url: process.env.REACT_APP_BACKEND_URL,
  method: "POST",
};

function NewUser() {
  return (
    <>
      <Header {...headerProps} />
      <UserForm {...userFormProps} />
      <Link to="/users">Back to users</Link>
    </>
  );
}

export default NewUser;
