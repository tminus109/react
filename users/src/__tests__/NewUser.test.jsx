import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import NewUser from "../pages/NewUser";

function renderContainer() {
  render(
    <MemoryRouter>
      <NewUser />
    </MemoryRouter>
  );
}

test("Snapshot test", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <NewUser />
      </MemoryRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("Smoke test", () => {
  const div = document.createElement("div");

  render(
    <MemoryRouter>
      <NewUser />
    </MemoryRouter>,
    div
  );
});

test("New User heading displayed properly", () => {
  renderContainer();

  const heading = screen.getByRole("heading", { name: /new user/i });

  expect(heading).toBeInTheDocument();
  expect(heading).toBeVisible();
  expect(heading.textContent).toBe("New User");
});

test("Add user btn displayed properly", () => {
  renderContainer();

  const btn = screen.getByRole("button", {
    name: /add user/i,
  });

  expect(btn).toBeInTheDocument();
  expect(btn).toBeVisible();
  expect(btn.textContent).toBe("Add user");
});

test("Form input labels displayed properly", () => {
  renderContainer();

  const firstNameLbl = screen.getByText(/first name:/i);
  const lastNameLbl = screen.getByText(/last name:/i);
  const statusLbl = screen.getByText(/status:/i);

  expect(firstNameLbl.textContent).toBe("First name:");
  expect(lastNameLbl.textContent).toBe("Last name:");
  expect(statusLbl.textContent).toBe("Status:");
  expect(firstNameLbl).toBeVisible();
  expect(lastNameLbl).toBeVisible();
  expect(statusLbl).not.toBeVisible();
});

test("Form inputs displayed properly", () => {
  renderContainer();

  const firstNameInput = screen.getByRole("textbox", {
    name: /first name:/i,
  });

  const lastNameInput = screen.getByRole("textbox", {
    name: /last name:/i,
  });

  const statusInput = screen.getByLabelText("Status:");

  expect(firstNameInput.value).toBe("");
  expect(lastNameInput.value).toBe("");
  expect(statusInput.value).toBe("active");
  expect(firstNameInput).toBeVisible();
  expect(lastNameInput).toBeVisible();
  expect(statusInput).not.toBeVisible();
});

test(`"Back to users" link is in the document`, () => {
  renderContainer();

  const link = screen.getByRole("link", { name: /back to users/i });

  expect(link).toBeInTheDocument();
  expect(link).toBeVisible();
  expect(link.textContent).toBe("Back to users");
});
