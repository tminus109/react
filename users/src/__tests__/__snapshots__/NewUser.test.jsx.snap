// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
Array [
  <div>
    <header>
      <h1>
        New User
      </h1>
    </header>
  </div>,
  <div>
    <fieldset>
      <legend>
        Add new user
      </legend>
      <form
        onSubmit={[Function]}
      >
        <ul>
          <li>
            <label
              htmlFor="first_name"
              id="first-name-lbl"
            >
              First name:
            </label>
          </li>
          <li>
            <input
              autoFocus={true}
              id="first_name"
              onChange={[Function]}
              placeholder="First name..."
              type="text"
              value=""
            />
          </li>
          <li>
            <label
              htmlFor="last_name"
              id="last-name-lbl"
            >
              Last name:
            </label>
          </li>
          <li>
            <input
              id="last_name"
              onChange={[Function]}
              placeholder="Last name..."
              type="text"
              value=""
            />
          </li>
          <li
            hidden={true}
          >
            <label
              htmlFor="status"
            >
              Status:
            </label>
          </li>
          <li
            hidden={true}
          >
            <input
              id="status"
              onChange={[Function]}
              type="text"
              value="active"
            />
          </li>
          <li>
            <div>
              <button
                className="button"
              >
                Add user
              </button>
            </div>
          </li>
        </ul>
      </form>
    </fieldset>
  </div>,
  <a
    href="/users"
    onClick={[Function]}
  >
    Back to users
  </a>,
]
`;
